# **Waze User Churn: Data Overview**

### Identify data types and compile summary information

This notebook provides an initial overview of the Waze churn dataset, including data types, missing values, and basic behavior summaries for churned vs. retained users. The goal is to understand who is in the dataset and whether early patterns in driving behavior align with churn outcomes.

### **Importing and Loading Data**



In [1]:
# Import packages for data manipulation
import pandas as pd
import numpy as np

# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

#### **Data summary information**
The first few rows and basic DataFrame metadata help confirm schema, data types, and overall shape.

In [2]:
df.head(10)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


There are 14,999 rows and 13 columns, with a mix of float, integer, and object (string) variables. The dataset has 700 missing values in the `label` column, while all other fields are fully populated.


### **Inspecting missing values and summary statistics**

To check whether rows with missing churn labels differ systematically from complete cases, summary statistics are compared for the two groups.

In [3]:
# Isolate rows with null values
df_null = df[df['label'].isnull()]
# Display summary stats of rows with null values
df_null.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [4]:
# Isolate rows without null values
df_nonnull = df[df['label'].notnull()]
# Display summary stats of rows without null values
df_nonnull.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

The summary statistics for rows with and without missing churn labels are very similar, suggesting no obvious pattern or systematic bias in where the label is missing.

### **Missing churn label by device type**

Next, the distribution of missing churn labels is examined by device type to check for device‑specific bias.

In [5]:
# Get count of null values by device
print(df_null.groupby('device').size())
df_null['device'].value_counts() 

device
Android    253
iPhone     447
dtype: int64


device
iPhone     447
Android    253
Name: count, dtype: int64

253 Android users and 447 iPhone users have missing churn labels.

In [6]:
# Calculate % of iPhone nulls and Android nulls
df_null['device'].value_counts('iPhone','Android')

device
iPhone     0.638571
Android    0.361429
Name: proportion, dtype: float64

In [7]:
# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize=True)

device
iPhone     0.644843
Android    0.355157
Name: proportion, dtype: float64

The percentage of missing labels for each device closely matches each device’s share in the full dataset, which supports the assumption that churn label missingness is approximately random with respect to device type.

### **Churn vs. retention distribution**

The churn label distribution indicates how imbalanced the classification problem is.

In [8]:
# Calculate counts of churned vs. retained
print(df['label'].value_counts())
print()
df['label'].value_counts('retained','churned')

label
retained    11763
churned      2536
Name: count, dtype: int64



label
retained    0.822645
churned     0.177355
Name: proportion, dtype: float64

Among users with known churn status, roughly 82% are retained and 18% are churned, indicating a moderately imbalanced classification task.


### **Median behavior by churn status**

Median values are used to summarize typical churned vs. retained users because several variables contain extreme values and skewed distributions. The median is more robust to outliers than the mean and better reflects a “typical” user in this context.


In [9]:
# Calculate median values of all numeric columns for churned and retained users
df.groupby('label').median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

Users who churned completed slightly more drives in the last month than retained users, but retained users used the app on more than twice as many days. Churned users also drove more total distance and spent more time driving, suggesting fewer but longer and more intensive driving days.

### **Distance per drive and per driving day**

To better understand intensity of use, efficiency metrics are derived for kilometers per drive and per driving day, segmented by churn status.

In [10]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Group by `label`, calculate the median km per drive
df.groupby('label')['km_per_drive'].median()

# Exemplar median table
median_km_per_drive = df.groupby('label').median(numeric_only=True)[['km_per_drive']]
median_km_per_drive

km_per_drive
label                 
churned      74.109416
retained     75.014702

The median retained user drives slightly farther per trip than the median churned user, but this gap is smaller than the difference in overall distance per driving day.

In [11]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days']

# Group by `label`, calculate the median km per driving day
df.groupby('label')['km_per_driving_day'].median()

label
churned     697.541999
retained    289.549333
Name: km_per_driving_day, dtype: float64

In [12]:
# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Group by `label`, calculate the median drives per driving day
df.groupby('label')['drives_per_driving_day'].median()

label
churned     10.0000
retained     4.0625
Name: drives_per_driving_day, dtype: float64

Churned users drive substantially more kilometers and complete more drives per driving day than retained users, indicating a segment of very high‑intensity drivers. This group likely represents professional or long‑haul users whose needs may differ from typical commuters and who may churn if the app does not fully support their workflows.

### **Device mix and churn**

Device distribution is examined within churned and retained groups to assess whether churn is associated with iPhone vs. Android usage.

In [13]:
# For each label, calculate the number of Android users and iPhone users
grouped = df.groupby(['label', 'device']).size()
grouped

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [14]:
# For each label, calculate the ratio of Android to iPhone users
print('churned', grouped.loc[('churned', 'Android')] / grouped.loc[('churned', 'iPhone')])
print('retained', grouped.loc[('retained', 'Android')] / grouped.loc[('retained', 'iPhone')])

churned 0.5416413373860183
retained 0.5518469656992084


The Android‑to‑iPhone ratio is very similar for churned and retained users and closely matches the overall device mix (approximately 64% iPhone and 36% Android), suggesting no meaningful difference in churn rate between device types.

## **Initial insights and next steps**

- The dataset contains 14,999 users with 700 missing churn labels, all in the `label` column; missingness appears roughly random and not concentrated in any specific device segment.
- The churn label is moderately imbalanced (about 18% churn), which will need to be considered when training and evaluating predictive models. 
- Churned users tend to drive more total distance and complete more drives per driving day than retained users, pointing to a high‑intensity segment that may have distinct needs and churn drivers.
- Device type (iPhone vs. Android) does not show an appreciable difference in churn behavior in this dataset.

These findings motivate deeper exploratory analysis and feature engineering in subsequent notebooks, with particular attention to heavy‑usage “super‑drivers” and how their behavior evolves leading up to churn.